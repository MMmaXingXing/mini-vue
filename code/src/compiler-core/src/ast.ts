import { CREATE_ELEMENT_VNODE } from "./runtimeHelpers";

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION
}

export const createVNodeCall = (context, tag, props, children) => {
  context.helper(CREATE_ELEMENT_VNODE);

  return {
    type: NodeTypes.ELEMENT,
    tag: tag,
    props: props,
    children: children
  };
};
