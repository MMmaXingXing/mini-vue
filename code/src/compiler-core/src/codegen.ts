import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING
} from "./runtimeHelpers";

export const generate = (ast) => {
  const context = createCodegenContext();
  const { push } = context;

  genFunctionPreamble(context, ast);
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(",");

  push(`function ${functionName}(${signature}){`);
  push(`return `);
  genNode(ast.codegenNode, context);
  push("}");
  return {
    code: context.code
  };
};
const genFunctionPreamble = (context, ast) => {
  const { push } = context;
  const vueBinging = "Vue";
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelper)} } = ${vueBinging}`);
  }
  push("\n");
  push("return ");
};

const createCodegenContext = () => {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    }
  };
  return context;
};

const genNode = (node, context) => {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(context, node);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
};

const genCompoundExpression = (node, context) => {
  const { push } = context;
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
};

const genElement = (node, context) => {
  const { push, helper } = context;
  const { tag, children, props } = node;
  // const child = children[0];

  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // genNode(child, context);
  // 元素节点最顶层只会有一层，因此
  // for (let i = 0; i < children.length; i++) {
  //   // 新增复合节点类型 compound
  //   const child = children[i];
  //   genNode(child, context);
  // }
  genNodeList(genNullable([tag, props, children]), context);
  // genNode(children, context);
  push(")");
};

const genNodeList = (nodes, context) => {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }

    if (i < nodes.length - 1) {
      push(", ");
    }
  }
};

const genNullable = (args: any) => {
  return args.map((arg) => arg || "null");
};

const genExpression = (node, context) => {
  const { push } = context;
  push(`${node.content}`);
};

const genInterpolation = (node, context) => {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
};

const genText = (context, node) => {
  const { push } = context;
  push(`'${node.content}'`);
};
