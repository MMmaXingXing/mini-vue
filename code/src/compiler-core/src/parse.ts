import { NodeTypes } from "./ast";

const enum TagType {
  START,
  END
}

export const baseParse = (content: string) => {
  // 创建一个全局的上下文对象
  const context = createParseContext(content);
  return createRoot(parseChildren(context));
};

const parseChildren = (context) => {
  const nodes: any[] = [];

  let node;
  const s = context.source;
  if (s.startsWith("{{")) {
    node = parseInterpolation(context);
  } else if (s[0] === "<") {
    if (/[a-z]/i.test(s[1])) {
      console.log("parse elemet");
      node = parseElement(context);
    }
  }

  nodes.push(node);

  return nodes;
};

const parseElement = (context) => {
  // Implement
  // 1. 解析tag
  const element = parseTag(context, TagType.START);
  parseTag(context, TagType.END);
  return element;
};

const parseTag = (context, type: TagType) => {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];

  // 2. 删除处理完的代码
  advanceBy(context, match[0].length);
  console.log(context.source);
  advanceBy(context, 1);
  if (TagType.END) return;
  return {
    type: NodeTypes.ELEMENT,
    tag
  };
};

const parseInterpolation = (context) => {
  // {{message}}
  // 关键字变化点，抽取出来使程序解耦
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    closeDelimiter.length
  );

  // 前两个数据可以将前两个字符串干掉来进行一字符的推进
  advanceBy(context, openDelimiter.length);

  const rowContentLength = closeIndex - openDelimiter.length;
  const rawContent = context.source.slice(0, rowContentLength);
  const content = rawContent.trim();

  // 获取之后删除
  advanceBy(context, rowContentLength + closeDelimiter.length);
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content
    }
  };
};

const advanceBy = (context: any, length: number) => {
  context.source = context.source.slice(length);
};

const createRoot = (children) => {
  return {
    children
  };
};

const createParseContext = (content: string) => {
  return {
    source: content
  };
};
