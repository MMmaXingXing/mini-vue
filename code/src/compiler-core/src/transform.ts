import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export const transform = (root, options = {}) => {
  const context = createTransformContext(root, options);
  // 1. 遍历 - 深度优先搜索
  traverseNode(root, context);
  // 2. 修改 text content

  // root.codegenNode
  createRootCodgen(root);

  root.helpers = [...context.helpers.keys()];
};

const createRootCodgen = (root) => {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT && child.codegenNode) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = child;
  }
};

const createTransformContext = (root, options) => {
  // 创建全局上下文对象来存储我们传入的数据
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    }
  };

  return context;
};

const traverseNode = (node, context) => {
  // 实现深度优先搜索
  // 1. element
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) exitFns.push(onExit);
  }

  // 判断类型
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      // 判断是不是插值 balabala
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
};

const traverseChildren = (node, context) => {
  const children = node.children;

  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
};
