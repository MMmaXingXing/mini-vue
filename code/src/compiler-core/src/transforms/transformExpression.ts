import { NodeTypes } from "../ast";

// 触发时包含在插值类型中
export const transformExpression = (node) => {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
};
function processExpression(node: any) {
  node.content = "_ctx." + node.content;
  return node;
}
