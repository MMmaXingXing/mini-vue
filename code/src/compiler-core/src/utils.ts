import { NodeTypes } from "./ast";

export const isText = (node) => {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;
};
