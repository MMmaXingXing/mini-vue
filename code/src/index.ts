// mini-vue 出口
export * from "./runtime-dom";
export * from "./reactivity";

import { baseCompile } from "./compiler-core/src";
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";

// render
const compileToFunction = (template) => {
  const { code } = baseCompile(template);

  const render = new Function("Vue", code)(runtimeDom);
  return render;
};

registerRuntimeCompiler(compileToFunction);
