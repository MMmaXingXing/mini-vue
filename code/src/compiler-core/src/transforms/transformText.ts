import { NodeTypes } from "../ast";
import { isText } from "../utils";

export const transformText = (node) => {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      let currentContainer;
      const { children } = node;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];

        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextChild = children[j];
            if (isText(nextChild)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child]
                };
              }

              currentContainer.children.push(" + ");
              currentContainer.children.push(nextChild);
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    };
  }
};
