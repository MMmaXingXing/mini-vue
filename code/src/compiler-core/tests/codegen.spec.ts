import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");

    transform(ast);
    const { code } = generate(ast);

    // 快照 （string）
    // 1. 抓bug
    // 2. 有意

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression]
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div></div>");
    transform(ast, {
      nodeTransforms: [transformElement]
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element-all", () => {
    const ast = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText]
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
