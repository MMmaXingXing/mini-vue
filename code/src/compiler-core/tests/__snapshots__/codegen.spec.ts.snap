// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"import { createElementVNode:_createElementVNode } from Vue
return function render(_ctx,_cache){return _createElementVNode(\\"div\\", null, null)}"
`;

exports[`codegen element-all 1`] = `
"import { toDisplayString:_toDisplayString,createElementVNode:_createElementVNode } from Vue
return function render(_ctx,_cache){return _createElementVNode(\\"div\\", null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interpolation 1`] = `
"import { toDisplayString:_toDisplayString } from Vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx,_cache){return 'hi'}"
`;
